### PR TITLE
 added build-arg flag to pass build time argument

### PR DESCRIPTION
**Pull request by s8sg**
Signed-off-by: Swarvanu Sengupta <swarvanusg@gmail.com>

## Description
`--build-arg` flag is added for `faas-cli build` command which will provide build time argument to underlying `docker build` command with existing build argument ( `http-proxy` & `https-proxy`)
build arguments can be specified as `faas-cli build --build-arg CAPTURE_THIS=my-value ...`
which would eventually call
```
docker build --build-argument http-proxy=<> --build-argument https-proxy=<> --build-argument CAPTURE_THIS =my-value ...
```
The `Dockerfile` must have `ARG` to accept the `build-arg`
```dockerfile
ARG CAPTURE_THIS
``` 

## Motivation and Context
[Issue-362](https://github.com/openfaas/faas-cli/issues/362), [Issue-262](https://github.com/openfaas/faas-cli/issues/262) and [Issue-355](https://github.com/openfaas/faas-cli/issues/355) provides the background of the same

This fix issue: [Issue-363](https://github.com/openfaas/faas-cli/issues/363)
- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?
Testing have been done manually with the provided Dockerfile 
Dockerfile
```dockerfile
rm stack.yml # no conflicts
faas new --lang dockerfile argstest

cat > ./argstest/Dockerfile
FROM alpine:3.6
# 1. Use any image as your base image, or "scratch"
# 2. Add fwatchdog binary via https://github.com/openfaas/faas/releases/
# 3. Then set fprocess to the process you want to invoke per request - i.e. "cat" or"my_binary"

#ADD https://github.com/openfaas/faas/releases/download/0.7.8/fwatchdog /usr/bin
#RUN chmod +x /usr/bin/fwatchdog

RUN apk --no-cache add curl \
    && echo "Pulling watchdog binary from Github." \
    && curl -sSL https://github.com/openfaas/faas/releases/download/0.7.8/fwatchdog > /usr/bin/fwatchdog \
    && chmod +x /usr/bin/fwatchdog \
    && apk del curl --no-cache

ARG CAPTURE_THIS

RUN echo ${CAPTURE_THIS} | tee /captured.txt
ENV fprocess="cat /captured.txt"
# Set to true to see request in function logs
ENV write_debug="false"

HEALTHCHECK --interval=5s CMD [ -e /tmp/.lock ] || exit 1
CMD [ "fwatchdog" ]
```
Build:
```bash
go build && ./faas-cli build -f ./argtest.yml --build-arg CAPTURE_THIS=my-value --no-cache
```
Check the data was populated:
```bash
docker run -ti argtest cat /captured.txt
my-value
```

## Types of changes

- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] All new and existing tests passed.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by s8sg** [comment]
I just realized its a duplicate PR

File: commands/build.go [hunk 1]
```diff
@@ -165,3 +171,12 @@ func PullTemplates(templateURL string) error {
 	}
 	return err
 }
+
+func compileArg(argOpts []string) (map[string]string, error) {
+	arguments, err := parseMap(argOpts, "build-arg")
```
Comment (alexellis) [inline commands/build.go]: Interesting re-use of this function. I added a new one in my version, but maybe you can send a PR to the master now and remove my newly added method that parses the values?

**Comment by alexellis** [comment]
Thank you so much for working on this @s8sg I had no idea you were raising this PR.

Maybe next time you could also add a note to the issue to say you're taking it on?

Given how quickly you were able to do this I'm sure we can find you something else to do too. How does that sound?

File: commands/build.go (hunk 1 above)
Comment (s8sg) [reply commands/build.go]: @alexellis sure I'm on it :)

**Comment by s8sg** [comment]
@alexellis Sound great. I would love to contribute more :)

**Pull request closed without merging**